File: src/BaseHmiTypes/Scripts/HmiAction.cs
using BaseHmiTypes.Screens.Base;

namespace BaseHmiTypes.Scripts;

public enum HmiActionKind
{
    CommandList,
    TextScript,
    OpenResource
}

public abstract class HmiAction
{
    public abstract HmiActionKind Kind { get; }
}

public sealed class HmiCommandListAction : HmiAction
{
    public override HmiActionKind Kind => HmiActionKind.CommandList;

    public IList<HmiScriptCommand> Commands { get; } = new List<HmiScriptCommand>();
}

public sealed class HmiTextScriptAction : HmiAction
{
    public override HmiActionKind Kind => HmiActionKind.TextScript;

    public HmiScriptLanguage Language { get; set; } = HmiScriptLanguage.Unknown;

    public string? SourceCode { get; set; }
}

/// <summary>
/// Describes an inert request to open a URL or file resource.
/// </summary>
public sealed class HmiOpenResourceAction : HmiAction
{
    public override HmiActionKind Kind => HmiActionKind.OpenResource;

    public HmiProperty<string>? Address { get; set; }
}

File: typescript/base-hmi-types/src/scripts/HmiAction.ts
import {
  HmiScriptLanguage,
  type HmiProperty,
} from "../screens/base/HmiProperty.js";
import { HmiScriptCommand } from "./HmiScriptCommand.js";

export enum HmiActionKind {
  CommandList = "CommandList",
  TextScript = "TextScript",
  OpenResource = "OpenResource",
  OleVerb = "OleVerb",
}

export abstract class HmiAction {
  abstract readonly kind: HmiActionKind;
}

export class HmiCommandListAction extends HmiAction {
  readonly kind = HmiActionKind.CommandList;
  readonly commands: HmiScriptCommand[] = [];
}

export class HmiTextScriptAction extends HmiAction {
  readonly kind = HmiActionKind.TextScript;
  language: HmiScriptLanguage = HmiScriptLanguage.Unknown;
  sourceCode?: string;
}

/** Describes an inert request to open a URL or file resource. */
export class HmiOpenResourceAction extends HmiAction {
  readonly kind = HmiActionKind.OpenResource;
  address?: HmiProperty<string>;
}

/** Describes an inert request to invoke a verb on an embedded OLE object. */
export class HmiOleVerbAction extends HmiAction {
  readonly kind = HmiActionKind.OleVerb;
  verb?: string;
  securityCode?: string;
}
